source/backend/api/Services/DocumentActivityService.cs

```diff
@@ -0,0 +1,150 @@
+            this.documentActivityTemplateRepository = documentActivityTemplateRepository;
+            this.documentService = documentService;
+            this.mapper = mapper;
+        }
+
+        public IList<PimsActivityInstanceDocument> GetActivityDocuments(long activityId)
+        {
+            this.Logger.LogInformation("Retrieving PIMS document for single activity");
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentView);
+
+            return documentActivityRepository.GetAllByActivity(activityId);
+        }
+
+        public IList<PimsActivityTemplateDocument> GetActivityTemplateDocuments(long activityTemplateId)
+        {
+            this.Logger.LogInformation("Retrieving PIMS document for single activity template");
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentView);
+
+            return documentActivityTemplateRepository.GetAllByActivityTemplate(activityTemplateId);
+        }
+
+        public async Task<DocumentUploadRelationshipResponse> UploadActivityDocumentAsync(long activityId, DocumentUploadRequest uploadRequest)
+        {
+            this.Logger.LogInformation("Uploading document for single activity");
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentAdd);
+
+            DocumentUploadResponse uploadResult = await documentService.UploadDocumentAsync(uploadRequest);
+
+            DocumentUploadRelationshipResponse relationshipResponse = new DocumentUploadRelationshipResponse()
+            {
+                UploadResponse = uploadResult,
+            };
+
+            if (uploadResult.Document.Id != 0)
+            {
+                // Create the pims document activity relationship
+                PimsActivityInstanceDocument newActivityDocument = new PimsActivityInstanceDocument()
+                {
+                    ActivityInstanceId = activityId,
+                    DocumentId = uploadResult.Document.Id,
+                };
+                newActivityDocument = documentActivityRepository.Add(newActivityDocument);
+                documentActivityRepository.CommitTransaction();
+
+                relationshipResponse.DocumentRelationship = mapper.Map<DocumentRelationshipModel>(newActivityDocument);
+            }
+
+            return relationshipResponse;
+        }
+
+        public async Task<DocumentUploadRelationshipResponse> UploadActivityTemplateDocumentAsync(long activityTemplateId, DocumentUploadRequest uploadRequest)
+        {
+            this.Logger.LogInformation("Uploading document for single activity");
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentAdmin);
+
+            DocumentUploadResponse uploadResult = await documentService.UploadDocumentAsync(uploadRequest);
+
+            DocumentUploadRelationshipResponse relationshipResponse = new DocumentUploadRelationshipResponse()
+            {
+                UploadResponse = uploadResult,
+            };
+
+            if (uploadResult.Document.Id != 0)
```

Comment:
I consistently get an NPE here when uploading a CDOGS template document. The uploadResult.Document is null, only the uploadResult.DocumentExternalResult is populated.

Reply:
uhm that didn't show up on my side. I'll double check

Reply:
This appears to be because the repo is returning an error and this service is not handling the error appropriately.

Reply:
It worked on my side. Looks like your mayan api is not returning a good result. I will make this check more robust